File: downloader_lib/parse.py
"""HTML parsing helpers for Vimm's Lair downloader."""
import re
from typing import List, Dict, Optional
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse, parse_qs, urlencode, urlunparse
import requests

BASE_URL = "https://vimm.net"
DOWNLOAD_BASE = "https://dl2.vimm.net"

def parse_game_details(html_content: str) -> Dict[str, any]:
    """Parse game details (size, format, rating) from game page HTML."""
    soup = BeautifulSoup(html_content, 'html.parser')
    details = {}
    
    # Find size and format in the download form or nearby text
    size_match = re.search(r'([0-9.]+)\s*(MB|GB|KB)', html_content, re.IGNORECASE)
    if size_match:
        size_value = float(size_match.group(1))
        size_unit = size_match.group(2).upper()
        # Convert to bytes
        multipliers = {'KB': 1024, 'MB': 1024**2, 'GB': 1024**3}
        details['size_bytes'] = int(size_value * multipliers.get(size_unit, 1))
        details['size_display'] = f"{size_match.group(1)} {size_unit}"
    
    # Find format/extension from mediaId or file type indicators
    format_match = re.search(r'\.([a-z0-9]{2,5})\b', html_content, re.IGNORECASE)
    if format_match:
        details['extension'] = format_match.group(1).lower()
    
    # Find rating (star icons or score)
    rating_container = soup.find('div', class_=re.compile(r'rating|score', re.I))
    if rating_container:
        stars = len(rating_container.find_all('img', src=re.compile(r'star', re.I)))
        if stars > 0:
            details['rating'] = stars
    else:
        # Look for text-based rating
        rating_text = re.search(r'(\d+(\.\d+)?)\s*(?:out of|/)\s*\d+', html_content)
        if rating_text:
            details['rating'] = float(rating_text.group(1))
    
    return details

def parse_games_from_section(html_content: str, section: str) -> List[Dict[str, str]]:
    """Parse a list of games from the HTML of a section page."""
    games = []
    soup = BeautifulSoup(html_content, 'html.parser')
    table = soup.find('table', {'class': 'rounded centered cellpadding1 hovertable striped'})
    if not table:
        return []

    rows = table.find_all('tr')
    for row in rows:
        first_td = row.find('td')
        if first_td:
            link = first_td.find('a')
            if link:
                name = link.text.strip()
                href = link.get('href')
                if not href:
                    continue
                if isinstance(href, (list, tuple)):
                    href = href[0]
                href = str(href)
                game_id = href.split('/')[-1]
                page_url = BASE_URL + href
                games.append({
                    'name': name,
                    'page_url': page_url,
                    'game_id': game_id,
                    'section': section
                })
    return games

def resolve_download_form(html_content: str, session: requests.Session, game_page_url: str, game_id: str, logger) -> Optional[str]:
    """Find the download form and resolve the final download URL, handling POSTs."""
    soup = BeautifulSoup(html_content, 'html.parser')
    dl_form = soup.find(id='dl_form') or soup.find('form', attrs={'id': 'dl_form'})

    media_id = None
    if dl_form:
        media_input = dl_form.find('input', attrs={'name': 'mediaId'}) or dl_form.find('input', attrs={'name': re.compile(r'^mediaId$', re.I)})
        if media_input:
            media_id = media_input.get('value')

        action = (dl_form.get('action') or '').strip()
        params = {inp.get('name'): inp.get('value') for inp in dl_form.find_all('input') if inp.get('name') and inp.get('value') is not None}
        if media_id and 'mediaId' not in params:
            params['mediaId'] = media_id

        if action:
            action_url = urljoin(BASE_URL + '/', action)
            method = (dl_form.get('method') or 'get').lower()
            parsed = urlparse(action_url)
            q = parse_qs(parsed.query)
            for k, v in list(q.items()):
                if k not in params and isinstance(v, list) and v:
                    params[k] = v[-1]

            if method == 'get':
                new_q = urlencode(params, doseq=False)
                action_url = urlunparse((parsed.scheme, parsed.netloc, parsed.path, parsed.params, new_q, parsed.fragment))
                if logger: logger.info(f"Resolved download URL via form action (GET) for {game_id}: {action_url}")
                return action_url
            else: # POST
                # For POST forms, extract the mediaId and construct a direct download URL
                # The server expects a GET request to the download URL with cookies from visiting the game page
                media_id = params.get('mediaId')
                if media_id:
                    # Use the action URL's netloc (dl3.vimm.net, etc.) with mediaId as query param
                    download_url = f"{parsed.scheme}://{parsed.netloc}/?mediaId={media_id}"
                    # Include alt parameter if present
                    if 'alt' in params:
                        download_url += f"&alt={params['alt']}"
                    if logger: logger.info(f"Constructed download URL from POST form for {game_id}: {download_url}")
                    return download_url

    # Fallbacks if form parsing fails
    a = soup.find('a', href=re.compile(r'mediaId=', re.IGNORECASE))
    if a and a.get('href'):
        resolved = urljoin(BASE_URL + '/', a.get('href'))
        if logger: logger.info(f"Resolved download URL via anchor for {game_id}: {resolved}")
        return resolved

    if not media_id:
        alt = soup.find('input', attrs={'name': re.compile(r'^mediaId$', re.I)})
        media_id = alt.get('value') if alt else None
    if media_id:
        fallback = f"{DOWNLOAD_BASE}/?mediaId={media_id}"
        if logger: logger.info(f"Fallback constructed download URL for {game_id}: {fallback}")
        return fallback

    return None

File: src/downloader_lib/parse.ts
import { parse } from "node-html-parser";
import { URL } from "url";
import { ROM_EXTENSIONS } from "../utils/constants";

const BASE_URL = "https://vimm.net";
const DOWNLOAD_BASE = "https://dl2.vimm.net";

export type GameEntry = {
	name: string;
	page_url: string;
	game_id: string;
	section: string;
};

export function parseGamesFromSection(
	htmlContent: string,
	section: string
): GameEntry[] {
	const root = parse(htmlContent);
	const table = root.querySelector(
		"table.rounded.centered.cellpadding1.hovertable.striped"
	);
	if (!table) return [];
	const rows = table.querySelectorAll("tr");
	const games: GameEntry[] = [];
	rows.forEach((row) => {
		const firstTd = row.querySelector("td");
		if (!firstTd) return;
		const link = firstTd.querySelector("a");
		if (!link) return;
		const name = link.text.trim();
		let href = link.getAttribute("href") || "";
		if (!href) return;
		const gameId = href.split("/").pop() || "";
		const pageUrl = BASE_URL + href;
		games.push({ name, page_url: pageUrl, game_id: gameId, section });
	});
	return games;
}

export async function resolveDownloadForm(
	htmlContent: string,
	fetchFn: (input: RequestInfo, init?: RequestInit) => Promise<Response>,
	gamePageUrl: string,
	gameId: string,
	logger?: any
): Promise<string | null> {
	const root = parse(htmlContent);
	const dlForm =
		root.querySelector("#dl_form") || root.querySelector("form[id=dl_form]");
	let mediaId: string | null = null;

	if (dlForm) {
		const formMethod = (dlForm.getAttribute("method") || "GET").toUpperCase();
		const formAction = (dlForm.getAttribute("action") || "").trim();
		try {
			const inputsInfo: any[] = [];
			dlForm.querySelectorAll("input").forEach((inp) => {
				inputsInfo.push({
					name: inp.getAttribute("name"),
					value: inp.getAttribute("value"),
				});
			});
			if (logger)
				logger.info(
					`Found download form for ${gameId}: method=${formMethod}, action=${formAction}, inputs=${JSON.stringify(
						inputsInfo
					)}`
				);
		} catch (e) {
			if (logger)
				logger.exception?.(`Error logging download form for ${gameId}`);
		}

		const mediaInput =
			dlForm.querySelector("input[name=mediaId]") ||
			dlForm.querySelector("input[name^=mediaId i]");
		if (mediaInput) mediaId = mediaInput.getAttribute("value") || null;

		const action = formAction;
		const params: Record<string, string> = {};
		dlForm.querySelectorAll("input").forEach((inp) => {
			const n = inp.getAttribute("name");
			const v = inp.getAttribute("value");
			if (n && v != null) params[n] = v;
		});
		if (mediaId && !params.mediaId) params.mediaId = mediaId;

		if (action) {
			const actionUrl = new URL(action, BASE_URL + "/");
			const method = (dlForm.getAttribute("method") || "get").toLowerCase();

			if (method === "get") {
				const newQ = new URL(actionUrl.toString());
				Object.keys(params).forEach((k) => newQ.searchParams.set(k, params[k]));
				const out = newQ.toString();
				if (logger)
					logger.info?.(
						`Resolved download URL via form action (GET) for ${gameId}: ${out}`
					);
				return out;
			} else {
				try {
					// Build basic headers
					const baseHeaders: Record<string, string> = {
						Referer: gamePageUrl,
						Accept:
							"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
					};
					const resp = await fetchFn(actionUrl.toString(), {
						method: "POST",
						body: new URLSearchParams(params),
						headers: baseHeaders,
					});
					if (resp && resp.ok && resp.url) {
						if (logger)
							logger.info?.(
								`POST form resolved to URL for ${gameId}: ${resp.url}`
							);
						return resp.url;
					}

					// GET fallback
					const getUrl = new URL(actionUrl.toString());
					Object.keys(params).forEach((k) =>
						getUrl.searchParams.set(k, params[k])
					);
					if (logger)
						logger.warn?.(
							`POST form did not return usable URL for ${gameId} (status=${
								resp.status
							}); trying GET fallback ${getUrl.toString()}`
						);
					try {
						const respGet = await fetchFn(getUrl.toString(), {
							method: "GET",
							headers: baseHeaders,
						});
						if (respGet && respGet.ok) {
							if (logger)
								logger.info?.(
									`GET fallback returned URL for ${gameId}: ${respGet.url}`
								);
							return respGet.url || getUrl.toString();
						}
					} catch (e) {
						if (logger)
							logger.exception?.(
								`GET fallback failed for ${gameId} at ${getUrl.toString()}`
							);
					}

					// Try mirrors
					const mirrors = ["dl3.vimm.net", "dl2.vimm.net", "dl1.vimm.net"];
					const original = actionUrl.host;
					for (const m of mirrors) {
						if (m === original) continue;
						const mirrorUrl = new URL(actionUrl.toString());
						mirrorUrl.host = m;
						Object.keys(params).forEach((k) =>
							mirrorUrl.searchParams.set(k, params[k])
						);
						if (logger)
							logger.info?.(
								`Trying mirror ${mirrorUrl.toString()} for ${gameId}`
							);
						try {
							const r = await fetchFn(mirrorUrl.toString(), {
								method: "GET",
								headers: baseHeaders,
							});
							if (r && r.ok) {
								if (logger)
									logger.info?.(
										`Mirror ${m} returned URL for ${gameId}: ${r.url}`
									);
								return r.url || mirrorUrl.toString();
							}
						} catch (err) {
							if (logger)
								logger.exception?.(`Error trying mirror ${m} for ${gameId}`);
						}
					}

					if (logger)
						logger.warn?.(
							`All form submission attempts failed for ${gameId}; returning GET-like URL ${getUrl.toString()}`
						);
					return getUrl.toString();
				} catch (err) {
					const getUrl = new URL(actionUrl.toString());
					Object.keys(params).forEach((k) =>
						getUrl.searchParams.set(k, params[k])
					);
					if (logger)
						logger.exception?.(
							`POST form submission raised exception for ${gameId}, falling back to GET-like URL: ${getUrl.toString()}`
						);
					return getUrl.toString();
				}
			}
		}
	}

	// Fallbacks if form parsing fails — try anchors
	const candidates: [string, string | null][] = [];
	root.querySelectorAll("a").forEach((a) => {
		const href = a.getAttribute("href");
		if (!href) return;
		if (/mediaId=/i.test(href) || /\.(ciso|rvz|iso|gcm)($|\?)/i.test(href)) {
			const resolved = new URL(href, BASE_URL + "/").toString();
			const text = (a.text || "").toLowerCase();
			let ext: string | null = null;
			if (text.includes(".")) {
				const parts = text.split(".");
				ext = "." + parts[parts.length - 1];
			}
			if (!ext) {
				const m = href.match(/\.(ciso|rvz|iso|gcm)($|\?)/i);
				if (m) ext = m[0].toLowerCase().split("?")[0];
			}
			candidates.push([resolved, ext]);
		}
	});

	if (candidates.length) {
		const pref = [".ciso", ".rvz"];
		for (const p of pref) {
			for (const [url, ext] of candidates) {
				if (ext === p) {
					if (logger)
						logger.info?.(
							`Resolved download URL via preferred anchor for ${gameId}: ${url} (matched ${p})`
						);
					return url;
				}
			}
		}
		if (logger)
			logger.info?.(
				`Resolved download URL via anchor for ${gameId}: ${candidates[0][0]}`
			);
		return candidates[0][0];
	}

	if (!mediaId) {
		const alt = root.querySelector("input[name^=mediaId i]");
		mediaId = alt?.getAttribute("value") || null;
	}
	if (mediaId) {
		const fallback = `${DOWNLOAD_BASE}/?mediaId=${mediaId}`;
		if (logger)
			logger.info?.(
				`Fallback constructed download URL for ${gameId}: ${fallback}`
			);
		return fallback;
	}

	return null;
}
